pkg/rpc/rpc.go

```diff
@@ -126,21 +135,22 @@ func (r *RPC) Close() error {
 	return r.writer.Close()
 }
 
+// setErr sets the err encoder/decoder fails and closes errChannel
 func (r *RPC) setErr(err error) {
-	r.errMu.Lock()
-	defer r.errMu.Unlock()
-
-	if r.err == nil {
-		pendingReqMsg := fmt.Sprintf("cancelling %d pending request(s)", len(r.pendingRequests))
+	r.errOnce.Do(func() {
 		if r.requestedShutdown {
-			r.logger.Infof("Plugin shutdown triggered: %s", pendingReqMsg)
+			r.logger.Info("Plugin shutdown triggered")
 		} else {
-			r.logger.Warnf("Plugin terminated unexpectedly: %s", pendingReqMsg)
+			r.logger.Warn("Plugin terminated unexpectedly")
+		}
+
+		if reqCount := len(r.pendingRequests); reqCount > 0 {
+			r.logger.Infof("Cancelling %d pending request(s)", reqCount)
```

Comment:
How are those pending requests being canceled? I'm only seeing channel operations on `r.pendigRequests` in the `*RPC.processResponses` method, which, however, exits after an error is set,.

Reply:
The pending requests are actually dropped because the rpc instance can no longer be used. The channel plugin must be reinitialized.